### PR TITLE
Add lint for unused lifetimes (fixes #459)

### LINKED ISSUE #459: Warning about useless lifetime
Hello.
The following code does not give any warning from `clippy`:

```rust
fn test<'a>(a: i32) -> i32 {
    a
}

fn main() {
    test(10);
}
```

but I think it should, because the lifetime is not used.

Thanks to add this lint.

### PR DESCRIPTION
**Pull request by Manishearth**
r? @llogiq

### CODE REVIEW

File: src/lifetimes.rs [hunk 1]
```diff
@@ -61,6 +64,7 @@ fn check_fn_inner(cx: &LateContext, decl: &FnDecl, slf: Option<&ExplicitSelf>,
         span_lint(cx, NEEDLESS_LIFETIMES, span,
                   "explicit lifetimes given in parameter types where they could be elided");
     }
+    report_extra_lifetimes(cx, decl, &generics.lifetimes);
```
Comment (llogiq) [inline src/lifetimes.rs]: I'm not sure, but won't this report methods in trait impls where the lifetime is specified on the trait definition? Would it be legal to remove it while leaving it on the trait def? I'm on mobile, so I can't check right now. I'd really like if fixing one lint warning (as opposed to all of them) doesn't introduce compilation errors.

**Comment by llogiq** [comment]
Minor nit aside, this looks plenty useful.

File: src/lifetimes.rs (hunk 1 above)
Comment (Manishearth) [inline src/lifetimes.rs]: It only reports lifetimes defined in the function lt param list that aren't used in the argument/return list, not the other way around. Adding a test.

(forgot to `git add` the tests folder too, did that now)

**Comment by llogiq** [comment]
LGTM. Merging.

**Pull request merged**